### PR TITLE
Codechange: Remove last FOR_XXX macros

### PR DESCRIPTION
**Pull request by glx22**
## Motivation / Problem
I finally finished removing `FOR_XXX` macros.

(Now Intellisense is more usable)

## Description
- Replaced manually allocated `HeightMap::h` with a vector
- Used a span for `_sorted_cargo_specs` sub-range of standard `CargoSpec`
- Introduced an iterator for bitset (and used it)

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/cargotype.cpp [hunk 1]
```diff
@@ -198,11 +198,13 @@ void InitializeSortedCargoSpecs()
...
-	_sorted_standard_cargo_specs_size = 0;
+	uint8 sorted_standard_cargo_specs_size = 0;
 	for (const auto &cargo : _sorted_cargo_specs) {
 		if (cargo->classes & CC_SPECIAL) break;
-		_sorted_standard_cargo_specs_size++;
+		sorted_standard_cargo_specs_size++;
 		SetBit(_standard_cargo_mask, cargo->Index());
 	}
+
+	_sorted_standard_cargo_specs = { _sorted_cargo_specs.data(), sorted_standard_cargo_specs_size };
```
Comment (TrueBrain) [inline src/cargotype.cpp]: This is a vector, so you should just be able to convert that to a `span<>`? Not really sure why the size needed to be calculated first?

File: src/core/bitmath_func.hpp [hunk 2]
```diff
@@ -345,20 +345,55 @@ static inline T ROR(const T x, const uint8 n)
...
+	};
+
+	BitIterator(Tbitset bitset) : bitset(bitset) {}
+	Iterator begin() { return Iterator(this->bitset); }
+	Iterator end() { return Iterator(static_cast<Tbitset>(0)); }
+	bool empty() { return this->begin() == this->end(); }
+
+private:
+	Tbitset bitset;
+};
```
Comment (TrueBrain) [inline src/core/bitmath_func.hpp:371]: Sadly, I have not enough knowledge to rate this part of the PR in any way :D @LordAro , mind giving this a look? Tnx!

File: src/core/bitmath_func.hpp [hunk 3]
```diff
@@ -320,31 +320,6 @@ static inline T ROR(const T x, const uint8 n)
 	return (T)(x >> n | x << (sizeof(x) * 8 - n));
 }
 
-/**
```
Comment (TrueBrain) [inline src/core/bitmath_func.hpp:323]: bit weird commit message .. `codechange` .. `remove`... :D

Maybe something like:

`Cleanup: remove now unused FOR_EACH_SET_BIT_EX macro`
?

File: src/cargotype.cpp (hunk 1 above)
Comment (glx22) [reply src/cargotype.cpp]: Because we want a subset of the vector (most of the time it will be full vector though)

Comment (TrueBrain) [reply src/cargotype.cpp]: Ah, the `break` there .. right .. that is sneaky :D

I think this deserves a comment somewhere :P

File: src/core/bitmath_func.hpp (hunk 2 above)
Comment (LordAro) [reply src/core/bitmath_func.hpp:371]: Dunno much either. Part of me wonders whether std::bitset might be useable somehow (though may well require a fair bit more refactoring to do so)

Comment (rubidium42) [reply src/core/bitmath_func.hpp:371]: Though bitset doesn't have iterators (or rather begin/end), so can it iterate over them? With my naive check you can't, and I think that was the whole point of making this iterator.

File: src/core/bitmath_func.hpp [hunk 4]
```diff
@@ -345,20 +345,55 @@ static inline T ROR(const T x, const uint8 n)
...
- * @param bitset_value The value which we check for set bits.
- */
-#define FOR_EACH_SET_BIT(bitpos_var, bitset_value) FOR_EACH_SET_BIT_EX(uint, bitpos_var, uint, bitset_value)
+ /**
+ * Iterable ensemble of each set bit in a value.
+ * @tparam Tbitpos Type of the position variable.
+ * @tparam Tbitset Type of the bitset value.
+*/
+template <typename Tbitpos = uint, typename Tbitset = uint>
+struct BitIterator {
```
Comment (rubidium42) [inline src/core/bitmath_func.hpp]: The name of the class does not imply that it only considers set bits, nor that it actually returns the position and not be bit value.

**Comment by TrueBrain** [comment]
I have no more comments. @rubidium42 : this okay with you too?

PS: really cool finally these really old macros are goneeeeeeee. Nice job @glx22 :D

**Review by TrueBrain — APPROVED** [review]
I am going to assume the silence is a yes :)

**Pull request merged**